### PR TITLE
[WASM] Add minimal support for targeting wasm32-unknown-unknown-wasm. Forked #20684

### PR DESCRIPTION
**Pull request by kateinoigakukun**
This PR is based on https://github.com/apple/swift/pull/20684 which adds the minimal wasm support.

And I added some test case to ensure that `swiftc` accepts `wasm32-unknown-unknown-wasm` and `wasm64-unknown-unknown-wasm`.

This is my first time to write lit test case, so please let me know if there is a problem.

### CODE REVIEW

File: include/swift/Basic/LangOptions.h [hunk 1]
```diff
@@ -346,6 +346,8 @@ namespace swift {
                  Target.isPS4() || Target.isOSHaiku() ||
                  Target.getTriple().empty()) {
         major = minor = revision = 0;
+      } else if (Target.isOSBinFormatWasm()) {
```
Comment (compnerd) [inline include/swift/Basic/LangOptions.h]: This seems wrong and is creating a branch which is going to cause more incorrect checks in the future.  This should be part of the check in the previous condition: the os is `none` should set `major` and `minor` to 0.

File: lib/Basic/LangOptions.cpp [hunk 2]
```diff
@@ -230,6 +230,8 @@ std::pair<bool, bool> LangOptions::setTarget(llvm::Triple triple) {
     addPlatformConditionValue(PlatformConditionKind::OS, "PS4");
   else if (Target.isOSHaiku())
     addPlatformConditionValue(PlatformConditionKind::OS, "Haiku");
+  else if (Target.isOSBinFormatWasm())
```
Comment (compnerd) [inline lib/Basic/LangOptions.cpp]: I'm not sure that we want to introduce this at all.  There shouldn't be anything specific to WASM here, it is an object file format, not an OS.  We should be introducing a `none` OS (which is the canonical spelling for this field being unset).

File: lib/Basic/Platform.cpp [hunk 3]
```diff
@@ -137,6 +137,11 @@ static StringRef getPlatformNameForDarwin(const DarwinPlatformKind platform) {
 StringRef swift::getPlatformNameForTriple(const llvm::Triple &triple) {
   switch (triple.getOS()) {
   case llvm::Triple::UnknownOS:
+    // NOTE: WebAssembly doesn't yet have a defined OS convention in triples.
+    if (triple.isOSBinFormatWasm()) {
```
Comment (compnerd) [inline lib/Basic/Platform.cpp]: Once we check the triple for `none`, this is going to be handled properly.

File: lib/Driver/Driver.cpp [hunk 4]
```diff
@@ -266,6 +266,11 @@ Driver::buildToolChain(const llvm::opt::InputArgList &ArgList) {
   case llvm::Triple::Haiku:
     return llvm::make_unique<toolchains::GenericUnix>(*this, target);
   default:
+    // NOTE: WebAssembly doesn't yet have a defined OS convention in triples.
+    if (target.isOSBinFormatWasm()) {
```
Comment (compnerd) [inline lib/Driver/Driver.cpp]: This again should be handled within the `none` case for the triple.

**Review by compnerd — CHANGES_REQUESTED** [review]

File: lib/Basic/LangOptions.cpp (hunk 2 above)
Comment (kateinoigakukun) [reply lib/Basic/LangOptions.cpp]: Hmm, AFAIK `llvm::Triple` can't represent `none` OS you said.
You mean we should add `none` case to `llvm::Triple`?

Comment (kateinoigakukun) [reply lib/Basic/LangOptions.cpp]: > There shouldn't be anything specific to WASM here, it is an object file format, not an OS. 

I agree that and I think it's necessary to handle `isOSWASI` case. 

>  We should be introducing a none OS (which is the canonical spelling for this field being unset).

You mean we should add `none` enum case to `llvm::Triple`? What the difference of `UnknownOS`?

Comment (compnerd) [reply lib/Basic/LangOptions.cpp]: `none` is the canonical spelling for *no* OS, `unknown` is meant for the vendor field only.  There is no such thing as an unknown OS, you either have an OS or do not.  The compiler needs to know what the code is being generated for.

Comment (kateinoigakukun) [reply lib/Basic/LangOptions.cpp]: OK, I understood what `none` OS means. Again, you mean we should add `llvm::Triple::None`, right?

**Comment by karwa** [comment]
In the time since the precursor to this PR was opened, our fork of LLVM has advanced enough to the point that [it supports `wasi`](https://github.com/apple/llvm-project/blob/apple/master/llvm/include/llvm/ADT/Triple.h#L192). Rather than trying to support a "none" OS, why don't we switch to supporting the wasi triples (`wasm32-wasi` and `wasm64-wasi`) instead? That would seem to address @compnerd 's issues.

In platform-specific code, that's what you're going to be interested in, anyway:
```swift
#if os(wasi)
import WasiLibc // wasi's stripped-down libc

printf(...)

#endif
```

**Comment by kateinoigakukun** [comment]
I updated to use WASI as target OS

**Comment by karwa** [comment]
The original patch is now over a year old, then it finally became unarguably ready to merge, so of course _someone_ had to immediately go and refactor that part of code so it now needs rewriting 😅

One day we will finally catch up to the likes of C, Forth and Lua, and support WASM.

**Comment by kateinoigakukun** [comment]
Hi, is there anything that prevents this PR from being merged?

**Comment by karwa** [comment]
@kateinoigakukun i believe @compnerd  redid this after his changes and merged that version

**Pull request closed without merging**